lightning-c-bindings/README.md

```diff
@@ -0,0 +1,143 @@
+The wrapper crate and C/C++ Headers in this folder are auto-generated from the Rust-Lightning
+source by the c-bindings-gen crate contained in the source tree and
+[cbindgen](https://github.com/eqrion/cbindgen). They are intended to be used as a base for building
+language-specific bindings, and are thus incredibly low-level and may be difficult to work with
+directly.
+
+In other words, if you're reading this, you're either writing bindings for a new language, or
+you're in the wrong place - individual language bindings should come with their own documentation.
+
+LDK C Bindings
+===================
+
+The C bindings available at include/lightning.h require inclusion of include/rust_types.h first.
+
+All of the Rust-Lightning types are mapped into C equivalents which take a few forms, namely:
+
+ * Structs are mapped into a simple wrapper containing a pointer to the native Rust-Lightning
+   object and a flag to indicate whether the object is owned or only a reference. Such mappings
+   usually generate a `X_free` function which must be called to release the allocated resources.
+   Note that calling `X_free` will do nothing if the underlying pointer is NULL or if the reference
+   flag is set.
+
+   You MUST NOT create such wrapper structs manually, relying instead on constructors which have
+   been mapped from equivalent Rust constructors.
+
+   Note that, thanks to the is-reference flag, such structs effectively represent both `&RustThing`
+   and `RustThing`. Further, the same applies to `Option<RustThing>`, in which cases `inner` may be
+   set to NULL.
+
+   For example, this is the mapping of ChannelManager.
+   ```
+   typedef struct MUST_USE_STRUCT LDKChannelManager {
+      /** ... */
+      LDKlnChannelManager *inner;
+      bool _underlying_ref;
+   } LDKChannelManager;
+   ```
+
+ * Traits are mapped into a concrete struct containing a void pointer and a jump table listing the
```

Comment:
Do add that the void pointer is always named `this_arg`.

Reply:
Also, and I might be wrong here, but I think another phrasing that might help folks seeing this for the first time is saying that this is essentially an adapter pattern.

Reply:
How about ```
   This should remind you of a C++ vtable, only written out by hand and with the class only
   containing a pointer, instead of the regular class data.
```